homeassistant/helpers/template.py

```diff
@@ -800,42 +897,4 @@ def is_safe_attribute(self, obj, attr, value):
             super().is_safe_attribute(obj, attr, value)
 
 
-ENV = TemplateEnvironment()
-ENV.filters['round'] = forgiving_round
-ENV.filters['multiply'] = multiply
-ENV.filters['log'] = logarithm
-ENV.filters['sin'] = sine
-ENV.filters['cos'] = cosine
-ENV.filters['tan'] = tangent
-ENV.filters['sqrt'] = square_root
-ENV.filters['as_timestamp'] = forgiving_as_timestamp
-ENV.filters['timestamp_custom'] = timestamp_custom
-ENV.filters['timestamp_local'] = timestamp_local
-ENV.filters['timestamp_utc'] = timestamp_utc
-ENV.filters['is_defined'] = fail_when_undefined
-ENV.filters['max'] = max
-ENV.filters['min'] = min
-ENV.filters['random'] = random_every_time
-ENV.filters['base64_encode'] = base64_encode
-ENV.filters['base64_decode'] = base64_decode
-ENV.filters['ordinal'] = ordinal
-ENV.filters['regex_match'] = regex_match
-ENV.filters['regex_replace'] = regex_replace
-ENV.filters['regex_search'] = regex_search
-ENV.filters['regex_findall_index'] = regex_findall_index
-ENV.filters['bitwise_and'] = bitwise_and
-ENV.filters['bitwise_or'] = bitwise_or
-ENV.globals['log'] = logarithm
-ENV.globals['sin'] = sine
-ENV.globals['cos'] = cosine
-ENV.globals['tan'] = tangent
-ENV.globals['sqrt'] = square_root
-ENV.globals['pi'] = math.pi
-ENV.globals['tau'] = math.pi * 2
-ENV.globals['e'] = math.e
-ENV.globals['float'] = forgiving_float
-ENV.globals['now'] = dt_util.now
-ENV.globals['utcnow'] = dt_util.utcnow
-ENV.globals['as_timestamp'] = forgiving_as_timestamp
-ENV.globals['relative_time'] = dt_util.get_age
-ENV.globals['strptime'] = strptime
+_NO_HASS_ENV = TemplateEnvironment(None)
```

Comment:
When is this used? And should we cache one or should we just generate it on the fly if it's not a common occurrence

Reply:
When templates are loaded from the config file they're not yet attached to any hass object, and they get validated in `config_validation` before the hass object is available.

It does end up getting used quite a bit. Once the template is actually connected to hass then it's not required.